File: zig/2024/day1/main.zig
const std = @import("std");
const print = std.debug.print;
const file = @embedFile("input.txt");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var list_left = std.ArrayList(i32).init(allocator);
    defer list_left.deinit();

    var list_right = std.ArrayList(i32).init(allocator);
    defer list_right.deinit();

    // input here.
    var tok = std.mem.tokenize(u8, file, " \n");
    while (tok.next()) |t| {
        const int_left = try std.fmt.parseInt(i32, t, 10);
        const int_right = try std.fmt.parseInt(i32, tok.next().?, 10);

        try list_left.append(int_left);
        try list_right.append(int_right);
    }

    std.mem.sort(i32, list_left.items, {}, comptime std.sort.asc(i32));
    std.mem.sort(i32, list_right.items, {}, comptime std.sort.asc(i32));

    // part1: calculate distance
    var total_distance: u32 = 0;
    for (list_left.items, 0..) |_, i| {
        total_distance += @abs(list_left.items[i] - list_right.items[i]);
    }

    print("Part 1: {d}\n", .{total_distance});

    // part2: calculate similarity
    var total_similarity: i32 = 0;
    var count_similarity: i32 = 0;
    for (list_left.items) |num_left| {
        for (list_right.items) |num_right| {
            if (num_left == num_right) {
                count_similarity += 1;
            }
        }
        total_similarity += num_left * count_similarity;
        count_similarity = 0;
    }

    print("Part 2: {d}\n", .{total_similarity});
}

File: go/2024/day1/cmd/main.go
package main

import (
	"fmt"
	"helpers"
	"math"
	"sort"
	"strconv"
	"strings"
)

func main() {
	file := helpers.ReadFile("../input.txt")

	leftString, rightString := []string{}, []string{}

	// populate the left and right lists using puzzle input
	for _, line := range file {
		lineSplit := strings.Split(line, "   ")
		leftString = append(leftString, lineSplit[0])
		rightString = append(rightString, lineSplit[1])
	}

	// convert slice of strings to slice of ints, need this to sort
	stringLen := len(leftString)
	leftInt, rightInt := make([]int, stringLen), make([]int, stringLen)

	for i := 0; i < stringLen; i++ {
		leftInt[i], _ = strconv.Atoi(leftString[i])
		rightInt[i], _ = strconv.Atoi(rightString[i])
	}

	// sort ints
	sort.Ints(leftInt)
	sort.Ints(rightInt)

	// part1: calculate distance
	totalDistance := 0.00
	for i := 0; i < len(leftInt); i++ {
		// why doesn't go have a built-in abs func for integers?
		// i don't want to write a func so i'll cast instead
		left := float64(leftInt[i])
		right := float64(rightInt[i])

		totalDistance += math.Abs(left - right)
	}

	fmt.Printf("Part 1: %d\n", int(totalDistance))

	// part2: calculate similarity
	totalSimilarity := 0
	countSimilarity := 0
	// iterate left list of integers
	for i := 0; i < len(leftInt); i++ {
		// count number of occurrences found in the right list
		for j := 0; j < len(rightInt); j++ {
			if leftInt[i] == rightInt[j] {
				countSimilarity += 1
			}
		}
		totalSimilarity += leftInt[i] * countSimilarity
		countSimilarity = 0
	}

	fmt.Printf("Part 2: %d\n", totalSimilarity)
}
